## ひろしまQuest2020#stayhome【球種予測部門】

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import lightgbm as lgb
import gc
%matplotlib inline
pd.set_option('display.max_Columns', 100)

In [2]:
ALL_MERGE = 'intermediate/all_merge_6.f'

In [3]:
all_pitch = pd.read_feather(ALL_MERGE)
all_pitch.shape

(778767, 222)

### 不要な列を削除

In [4]:
all_pitch.drop(
    columns=[
        '年度', 
        '試合ID', 
        'ホームチームID', 'アウェイチームID', 
        '投手ID', '投手チームID', 
        '打者ID', '打者チームID', 
        'プレイ前走者状況', 
        '捕手ID', 
        'opening_date', 'game_date',
        'start_time', 'game_time', 'elapsed_time'
    ], inplace=True)

### Rename

In [5]:
all_pitch.rename(columns={
    'データ内連番': 'No',
    '投球位置区域': 'pitch_area',
    '試合内連番': 'No_in_game',
    '試合内投球数': 'pitch_cnt_in_game',
    'イニング': 'inning',
    'イニング内打席数': 'bat_cnt_in_inning',
    '打席内投球数': 'pitch_cnt_in_bat',
    '投手登板順': 'pitch_order',
    '投手試合内対戦打者数': 'player_cnt_in_game',
    '投手試合内投球数': 'pitcher_cnt_in_game',
    '投手イニング内投球数': 'pitcher_cnt_in_inning',
    '打者打順': 'bat_order',
    '打者試合内打席数': 'bat_cnt_in_game',
    'プレイ前ホームチーム得点数': 'home_point',
    'プレイ前アウェイチーム得点数': 'away_point',
    'プレイ前アウト数': 'out_cnt',
    'プレイ前ボール数': 'ball_cnt',
    'プレイ前ストライク数': 'strike_cnt',
}, inplace=True)

In [6]:
print(all_pitch.shape)
all_pitch.head()

(778767, 207)


No  ball  pitch_area  No_in_game  pitch_cnt_in_game  inning  \
0   0   0.0         0.0           1                  1       1   
1   1   0.0         8.0           2                  2       1   
2   2   0.0         5.0           3                  3       1   
3   3   0.0        12.0           4                  4       1   
4   4   0.0         8.0           5                  5       1   

   bat_cnt_in_inning  pitch_cnt_in_bat  pitch_order  player_cnt_in_game  \
0                  1                 1            1                   1   
1                  1                 2            1                   1   
2                  1                 3            1                   1   
3                  2                 1            1                   2   
4                  2                 2            1                   2   

   pitcher_cnt_in_game  pitcher_cnt_in_inning  bat_order  bat_cnt_in_game  \
0                    1                      1          1                1   
1                    2                      2          1                1   
2                    3                      3          1                1   
3                    4                      4          2                1   
4                    5                      5          2                1   

   home_point  away_point  out_cnt  ball_cnt  strike_cnt  first  second  \
0           0           0        0         0           0      0       0   
1           0           0        0         0           1      0       0   
2           0           0        0         0           2      0       0   
3           0           0        1         0           0      0       0   
4           0           0        1         0           1      0       0   

   third  base_cnt  pitch_LR  bat_LR  top_bot  role  pos_pit  \
0      0         0       0.0     1.0        1     1        0   
1      0         0       0.0     1.0        1     1        0   
2      0         0       0.0     1.0        1     1        0   
3      0         0       0.0     0.0        1     1        0   
4      0         0       0.0     0.0        1     1        0   

   date_from_opening  elapsed_min  nighter  ce-pa  league  home  point_diff  \
0                  0     0.000000        1      0       1     1           0   
1                  0     0.200000        1      0       1     1           0   
2                  0     0.383333        1      0       1     1           0   
3                  0     1.016667        1      0       1     1           0   
4                  0     1.200000        1      0       1     1           0   

   runner_23  bant  squize  cleanup  pinch  fourball  savepoint  ball_cnt_0-0  \
0          0     0       0        1      0         0          0             1   
1          0     0       0        1      0         0          0             0   
2          0     0       0        1      0         0          0             0   
3          0     0       0        1      0         0          0             1   
4          0     0       0        1      0         0          0             0   

   ball_cnt_0-1  ball_cnt_0-2  ball_cnt_0-3  ball_cnt_1-0  ball_cnt_1-1  \
0             0             0             0             0             0   
1             0             0             0             1             0   
2             0             0             0             0             0   
3             0             0             0             0             0   
4             0             0             0             1             0   

   ball_cnt_1-2  ball_cnt_1-3  ...  pit_batter_per_inning  \
0             0             0  ...                    NaN   
1             0             0  ...                    NaN   
2             0             0  ...                    NaN   
3             0             0  ...                    NaN   
4             0             0  ...                    NaN   

   pit_ball_per_inning  pit_ball_per_batter  foreigner  batter_cnt  \
0     

### train
- 行数: 257117

In [7]:
train = all_pitch.dropna(subset=['pitch_area'])
train.shape

(257117, 207)

### test
- 行数: 521650

In [8]:
test = all_pitch[all_pitch['pitch_area'].isnull()]
test.shape

(521650, 207)

In [9]:
del all_pitch
gc.collect()

60

In [10]:
train_d = train.drop([
    'No', 
    'pitch_area', 
    'ball'
], axis=1)

In [11]:
test_d = test.drop([
    'No', 
    'pitch_area', 
    'ball'
], axis=1)

## Dataset作成

In [12]:
lgb_train = lgb.Dataset(train_d, train['ball'])

### LGBM学習
- multiclass : クラス数=8
- tuning結果のパラメータを使用
- 'learning_rate' : 0.1

In [13]:
%%time
lgb_param = {
        'objective' : 'multiclass',
        'boosting_type': 'gbdt',
        'metric' : 'multi_logloss',
        'num_class' : 8,
        'seed' : 0,
        'learning_rate' : 0.1,
        'lambda_l1': 6.9923570049658075, 
        'lambda_l2': 0.002378623984798833, 
        'num_leaves': 18, 
        'feature_fraction': 0.45199999999999996, 
        'bagging_fraction': 0.9799724836460725, 
        'bagging_freq': 4, 
        'min_child_samples': 20
    }

# 学習
cv_results = lgb.cv(lgb_param, lgb_train,
                    num_boost_round=15000,
                    early_stopping_rounds=100,
                    verbose_eval=100,
                    nfold=4)

num_boost_round = len(cv_results['multi_logloss-mean'])
print('Best num_boost_round:', num_boost_round)
print('Best CV score:', cv_results['multi_logloss-mean'][-1])

[100]	cv_agg's multi_logloss: 1.15208 + 0.00182535
[200]	cv_agg's multi_logloss: 1.13029 + 0.00192434
[300]	cv_agg's multi_logloss: 1.12288 + 0.00210615
[400]	cv_agg's multi_logloss: 1.11846 + 0.00234823
[500]	cv_agg's multi_logloss: 1.11515 + 0.00222199
[600]	cv_agg's multi_logloss: 1.11298 + 0.00222629
[700]	cv_agg's multi_logloss: 1.11134 + 0.00255114
[800]	cv_agg's multi_logloss: 1.1103 + 0.00255335
[900]	cv_agg's multi_logloss: 1.10946 + 0.00266343
[1000]	cv_agg's multi_logloss: 1.10902 + 0.0027239
[1100]	cv_agg's multi_logloss: 1.10873 + 0.00273811
[1200]	cv_agg's multi_logloss: 1.10873 + 0.00272691
Best num_boost_round: 1135
Best CV score: 1.1086837935965734
CPU times: user 1h 34min 51s, sys: 5.55 s, total: 1h 34min 57s
Wall time: 11min 54s


In [14]:
best_iter = int(num_boost_round * 1.1)

### 全体で再学習

In [15]:
%%time
lgb_model = lgb.train(lgb_param, lgb_train, num_boost_round=best_iter)

CPU times: user 30min 11s, sys: 1.29 s, total: 30min 12s
Wall time: 3min 46s


### Feature Importance

In [16]:
fi = lgb_model.feature_importance()
fn = lgb_model.feature_name()
df_feature_importance = pd.DataFrame({'feat_name':fn, 'feat_imp':fi})
df_feature_importance.sort_values('feat_imp', inplace=True)
df_feature_importance.tail(30)

feat_name  feat_imp
188              age_dif_p-b      1840
144        salary_x_year_bat      1899
15                strike_cnt      1909
63                  fork_pit      1935
107               height_bat      1960
6         player_cnt_in_game      2024
108               weight_bat      2040
190    salary_x_year_dif_p-b      2085
143          salary_year_bat      2088
119         bat_game_cnt_bat      2100
14                  ball_cnt      2155
198           salary_dif_b-c      2169
202    salary_x_year_dif_b-c      2241
31                point_diff      2243
61                slider_pit      2249
186           salary_dif_p-b      2251
118           batter_cnt_bat      2428
145                  bmi_bat      2440
120  batter_cnt_per_game_bat      2564
189      salary_year_dif_p-b      2682
201      salary_year_dif_b-c      2729
4           pitch_cnt_in_bat      2757
0                 No_in_game      3205
203              bmi_dif_b-c      3318
1          pitch_cnt_in_game      3321
191              bmi_dif_p-b      3345
26               elapsed_min      3680
8      pitcher_cnt_in_inning      3689
7        pitcher_cnt_in_game      4201
25         date_from_opening      5778

In [17]:
%%time
predict = lgb_model.predict(test_d, num_iteration = lgb_model.best_iteration)

CPU times: user 7min 13s, sys: 896 ms, total: 7min 14s
Wall time: 56 s


## 結果出力

In [18]:
submit = pd.DataFrame(predict)
submit.reset_index(inplace=True)
print(submit.shape)
submit.head()

(521650, 9)


index         0         1         2         3         4         5  \
0      0  0.300950  0.028488  0.147372  0.362830  0.003248  0.000017   
1      1  0.320805  0.084075  0.134868  0.384597  0.004379  0.000019   
2      2  0.289042  0.089023  0.158755  0.277100  0.080708  0.000078   
3      3  0.283279  0.097721  0.188402  0.308208  0.016394  0.000041   
4      4  0.217485  0.045145  0.272058  0.315432  0.011806  0.000035   

          6         7  
0  0.000016  0.157080  
1  0.000010  0.071247  
2  0.000020  0.105274  
3  0.000017  0.105939  
4  0.000016  0.138022

### 2017の平均
- ストレート           0.468254
- カーブ             0.074725
- スライダー           0.185806
- シュート            0.070633
- フォーク            0.083013
- チェンジアップ         0.054217
- シンカー            0.011361
- カットボール          0.051992

In [19]:
submit.mean()

index    260824.500000
0             0.519718
1             0.079965
2             0.182021
3             0.053803
4             0.070990
5             0.042196
6             0.007236
7             0.044071
dtype: float64

In [20]:
submit.to_csv('submit/lgbm_6.csv', header=False, index=False)

### 評価結果
#### 6 (207)
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- Best num_boost_round: 1135
- Best CV score: 1.1086837935965734
- Wall time: 11min 54s
- 評価結果  : 1.3999194

#### 5-tune (参考)
- 'num_leaves' : 18,
- 'learning_rate' : 0.05,
- Best num_boost_round: 2395
- Best CV score: 1.1103399384119979
- Wall time: 18min 29s (n1-standard-8)

#### 5-tune (135)
- 'num_leaves' : 18,
- 'learning_rate' : 0.1,
- CV score: 1.1112402566616455
- 評価結果  : 1.3991780
- CV Wall time: 27min 48s (local) -> 18min 24s (n1-standard-4) -> 9min 3s (n1-standard-8)

#### 5 (135)
- 'num_leaves' : 31,
- 'learning_rate' : 0.1,
- CV score: 1.1161825304353852
- 評価結果  : 1.4252286

#### 4 (128)
- CV score: 1.1158111693845418
- 評価結果  : 1.4299265

#### 3 (97)
- CV score: 1.1172290313076036
- 評価結果  : 1.4335806